client/web/src/enterprise/batches/preview/CreateUpdateBatchChangeAlert.tsx

```diff
@@ -1,99 +1,219 @@
+import { Action, DropdownButton } from '../DropdownButton'
+import { MultiSelectContext } from '../MultiSelectContext'
 
-import { createBatchChange, applyBatchChange } from './backend'
+import { applyBatchChange, createBatchChange, queryPublishableChangesetSpecs } from './backend'
+import { BatchChangePreviewContext } from './BatchChangePreviewContext'
 import styles from './CreateUpdateBatchChangeAlert.module.scss'
 
+export enum CreateUpdateBatchChangeAlertAction {
+    Apply,
+    PublishAll,
+    PublishSelected,
+    DraftAll,
+    DraftSelected,
+}
+
 export interface CreateUpdateBatchChangeAlertProps extends TelemetryProps {
     specID: string
     toBeArchived: number
     batchChange: BatchSpecFields['appliesToBatchChange']
+    showPublishUI: boolean
     viewerCanAdminister: boolean
     history: H.History
+    telemetryService: TelemetryService
 }
 
 export const CreateUpdateBatchChangeAlert: React.FunctionComponent<CreateUpdateBatchChangeAlertProps> = ({
     specID,
     toBeArchived,
     batchChange,
+    showPublishUI,
     viewerCanAdminister,
     history,
     telemetryService,
 }) => {
-    const batchChangeID = batchChange?.id
     const [isLoading, setIsLoading] = useState<boolean | Error>(false)
-    const [isRedesignEnabled] = useRedesignToggle()
 
-    const onApply = useCallback(async () => {
-        if (!confirm(`Are you sure you want to ${batchChangeID ? 'update' : 'create'} this batch change?`)) {
-            return
-        }
-        setIsLoading(true)
-        try {
-            const batchChange = batchChangeID
-                ? await applyBatchChange({ batchSpec: specID, batchChange: batchChangeID })
-                : await createBatchChange({ batchSpec: specID })
-
-            if (toBeArchived > 0) {
-                history.push(`${batchChange.url}?archivedCount=${toBeArchived}&archivedBy=${specID}`)
+    const { filters, pagination } = useContext(BatchChangePreviewContext)
+    const { selected } = useContext(MultiSelectContext)
+
+    const batchChangeID = batchChange?.id
+    const onApply = useCallback(
+        async (action: CreateUpdateBatchChangeAlertAction, setIsLoading: (loadingOrError: boolean | Error) => void) => {
+            if (!confirm(`Are you sure you want to ${batchChangeID ? 'update' : 'create'} this batch change?`)) {
+                return
+            }
+            setIsLoading(true)
+            try {
+                let publicationStates: ChangesetSpecPublicationStateInput[] | null = null
+                if (action !== CreateUpdateBatchChangeAlertAction.Apply) {
+                    const ids =
+                        selected === 'all'
+                            ? await queryPublishableChangesetSpecs({
+                                  batchSpec: specID,
+                                  ...filters,
+                                  ...pagination,
+                              }).toPromise()
+                            : [...selected]
```

Comment:
Sooo, after investigating what I described in my comment https://github.com/sourcegraph/sourcegraph/pull/22658#pullrequestreview-707145326 a bit, I think this is more what I would expect:

```suggestion
                    const queryAll =
                        selected === 'all' ||
                        action === CreateUpdateBatchChangeAlertAction.PublishAll ||
                        action === CreateUpdateBatchChangeAlertAction.DraftAll

                    const ids = queryAll
                        ? await queryPublishableChangesetSpecs({
                              batchSpec: specID,
                              ...filters,
                              ...pagination,
                          }).toPromise()
                        : [...selected]
```

That would make the `DraftAll` and `PublishAll` actions work as if the user hit the "Select all" button before.